keps/sig-cli/NNNN-kustomize-components.md

```diff
@@ -0,0 +1,308 @@
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+(**FIXME:** Explain what is the current situation and then roughly explain what this feature does)
+
+> This section is incredibly important for producing high quality user-focused
+> documentation such as release notes or a development roadmap.  It should be
+> possible to collect this information before implementation begins in order to
+> avoid requiring implementors to split their attention between writing release
+> notes and implementing the feature itself.  KEP editors, SIG Docs, and SIG PM
+> should help to ensure that the tone and content of the `Summary` section is
+> useful for a wide audience.
+>
+> A good summary is probably at least a paragraph in length.
+
+## Motivation
+
+(**FIXME:** Explain that we don't want to copy-paste stuff, list issues that wanted it at some form)
+
+> This section is for explicitly listing the motivation, goals and non-goals of
+> this KEP.  Describe why the change is important and the benefits to users.  The
+> motivation section can optionally provide links to [experience reports][] to
+> demonstrate the interest in a KEP within the wider Kubernetes community.
+>
+> [experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+
+### Goals
+
+(**FIXME:** Reusable configuration, anything else?)
+
+> List the specific goals of the KEP.  What is it trying to achieve?  How will we
+> know that this has succeeded?
+
+### Non-Goals
+
+(**FIXME:** What to add here?)
+
+> What is out of scope for this KEP?  Listing non-goals helps to focus discussion
+> and make progress.
+
+## Proposal
+
+(**FIXME:** High-level explanation of how it works)
+
+> This is where we get down to the specifics of what the proposal actually is.
+> This should have enough detail that reviewers can understand exactly what
+> you're proposing, but should not include things like API designs or
+> implementation.  The "Design Details" section below is for the real
+> nitty-gritty.
```

Comment:
This is what I propose for the **Proposal** section:

Create a new kustomization type called `Component`, and a new kustomization
field called `components`.

A kustomization that is marked as a `Component` has basically the same
capabilities as a normal kustomization. The main distinction is that they are
evaluated **after** the `resources` of an overlay have been accumulated, and
**on top** of them.  This means that:

* A component cannot be a top-level overlay. Instead, overlays are expected to
  be composed from multiple components, by adding them as a list in the
  `components` field.
* A component with transformers can transform the resources that an overlay has
  previously specified in the `resources` field. Components with patches do not
  have to include the target resource in their `resources` field.
* Multiple components can extend and transform the same set of resources
  **sequentially**. This is in contrast to overlays, which cannot alter the same
  base resources, because they clone them and extend them **in parallel**.

In order to create a component, a user can add the following to its
`kustomization.yaml`:

```yaml
apiVersion: kustomize.config.k8s.io/v1alpha1
kind: Component
```

In order to use a component, the user can refer to it in their
`kustomization.yaml` via the `components` field:

```yaml:
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

resources:
  - ../../base
  - resource1
  - resource2

components:
  - ../components/component1
  - ../components/component2
```

Reply:
Maybe say that a `Component` cannot be added to `resources:`, and `Kustomization` cannot be added to `components:`?

Reply:
Should we describe the alternative solution of allowing `Component` to be added to `resources:` here, or in the alternatives section?

Reply:
I'm not sure that we need the restriction to say that a component cannot be a top-level overlay - there is nothing in the implementation that would prevent that, and it might prove to be useful?

Reply:
Maybe say:
> The main distinction is that they are
evaluated after the resources of its **parent overlay (or component)** have been accumulated, and on top of them.

Reply:
> Maybe say that a Component cannot be added to resources:, and Kustomization cannot be added to components:?

Sure, makes sense.

> Should we describe the alternative solution of allowing Component to be added to resources: here, or in the alternatives section?

If I understand correctly, "Design details" is the section for discussing design choices in depth, so the rationale for the `components` field should be documented there.

> I'm not sure that we need the restriction to say that a component cannot be a top-level overlay - there is nothing in the implementation that would prevent that, and it might prove to be useful?

You're right, I'm introducing a restriction here for no reason. I'll remove this bullet point.

> Maybe say:
> 
> > The main distinction is that they are
> > evaluated after the resources of its parent overlay (or component) have been accumulated, and on top of them.

Ok.

Reply:
>The main distinction is that they are
evaluated after the resources of an overlay have been accumulated, and
on top of them. 

I would extend this to:

"The main distinction is that, once imported from another kustomization, they are evaluated after the resources of an overlay have been accumulated, and on top of them. "

Reply:
I like the "parent" term that @pgpx used, which implies the "import" you're saying above. However, I like that you refer to the parent as a kustomization, so I guess this is the best of both worlds:

"The main distinction is that they are evaluated **after** the resources of the parent kustomization (overlay or component) have been accumulated, and **on top** of them."